src/extension.ts

```diff
@@ -164,3 +169,67 @@ async function ensureRuntimeDependencies(extension: vscode.Extension<object>, lo
         return true;
     }
 }
+
+class DocumentLinkProvider implements vscode.DocumentLinkProvider {
+    private _client: LanguageClient;
+
+    constructor(client: LanguageClient) {
+        this._client = client;
+    }
+
+    // Provides the text document with the ranges for document links
+    provideDocumentLinks(document: vscode.TextDocument, token: CancellationToken): vscode.ProviderResult<vscode.DocumentLink[]> {
+        return this._findDocLinks(document);
+    }
+
+    // Adds the target uri to the document link
+    async resolveDocumentLink(link: vscode.DocumentLink, token: CancellationToken): Promise<vscode.DocumentLink> {
+        const document = vscode.window.activeTextEditor.document;
+        const text = document.getText(link.range);
+        link.target = await this._findTargetUri(document, text);
+        return link;
+    }
+
+    // Returns document links found within the current text document
+    private _findDocLinks(document: vscode.TextDocument): vscode.DocumentLink[] {
+        const docText = document.getText();
+        const docRoles = docText.match(/:doc:`.+?`/gs);
+
+        if (docRoles === null) return [];
+
+        let doclinks: vscode.DocumentLink[] = [];
+        let docRoleOffsetStart = -1; // Initiated to -1 to accommodate 0th index
+
+        // For every doc role found, find their respective target
+        docRoles.forEach(async (docRole, index) => {
+            docRoleOffsetStart = docText.indexOf(docRole, docRoleOffsetStart + 1);
+
+            // Find target in doc role
+            const targetMatches = docRole.match(/((?<=<).+(?=>))|((?<=`)\S+(?!>)(?=`))/);
+            const target = targetMatches[0];
+            const targetIndex = docRole.indexOf(target);
+
+            // Get range of the target within the scope of the whole text document
+            const targetOffsetStart = docRoleOffsetStart + targetIndex;
+            const targetOffsetEnd = targetOffsetStart + target.length;
+
+            doclinks.push({
+                range: new vscode.Range(
+                    document.positionAt(targetOffsetStart), 
+                    document.positionAt(targetOffsetEnd)
+                )
+            });
+        });
+
+        return doclinks;
+    }
+
+    // Returns the full uri given a target's name
+    private async _findTargetUri(document: vscode.TextDocument, target: string): Promise<vscode.Uri> {
+        return await this._client.sendRequest(
+            "textDocument/resolve", 
+            {filename: target, docpath: document.uri.path, resolve_type: "doc"}).then((file: string) => {
+                return vscode.Uri.file(file);
+        });
```

Comment:
What if resolve throws? Should we have a .catch() that logs the error?

Reply:
What would be an instance or example in which the resolve throws?

Reply:
Oh it absolutely shouldn't throw! I'm just paranoid 😄

Longer term, it might be valid to throw if the given path _can't_ be resolved.